File: bcc-apps/python/python_functions.c
/* Tracing Python functions */
#include <uapi/linux/ptrace.h>

struct data_t {
	char filename[128];
	char funcname[128];
	int lineno;
};
BPF_PERF_OUTPUT(events);

int print_functions(struct pt_regs *ctx)
{
	uint64_t argptr;
	struct data_t data = { };

	/* function__entry params are (str filename, str funcname, int lineno) */
	bpf_usdt_readarg(1, ctx, &argptr);
	bpf_probe_read_user(&data.filename, sizeof(data.filename),
			    (void *)argptr);
	bpf_usdt_readarg(2, ctx, &argptr);
	bpf_probe_read_user(&data.funcname, sizeof(data.funcname),
			    (void *)argptr);
	bpf_usdt_readarg(3, ctx, &data.lineno);

	// submit perf event.
	events.perf_submit(ctx, &data, sizeof(data));
	return 0;
};

File: bcc-apps/python/python_functions.py
#!/usr/bin/python3
#
# Tracing Python functions
import subprocess
from bcc import BPF, USDT
from bcc.utils import printb

# find the PID for "python3 -m http.server"
cmd = subprocess.Popen(
    ["pgrep", "-f", "http.server"], stdout=subprocess.PIPE, shell=False
).communicate()
if cmd[0]:
    pid = int(cmd[0].decode("ascii").strip())
else:
    print("ERROR: cannot find PID for python3 -m http.server")
    exit()

# load BPF program
u = USDT(pid=pid)
u.enable_probe(probe="function__entry", fn_name="print_functions")
b = BPF(src_file="python-functions.c", usdt_contexts=[u])


# callback for perf event
def print_event(cpu, data, size):
    event = b["events"].event(data)
    printb(b"%-9s %-6d %s" % (event.filename, event.lineno, event.funcname))


# print header
print("%-9s %-6s %s" % ("FILENAME", "LINENO", "FUNCTION"))

# loop with callback to print_event
b["events"].open_perf_buffer(print_event)
while 1:
    try:
        b.perf_buffer_poll()
    except KeyboardInterrupt:
        exit()
